### PR TITLE
feat: add `InvokeConstructor2` support

### PR DESCRIPTION
**Pull request by chanattan**
In continuation of #7834.

### CODE REVIEW

**Comment by chanattan** [comment]
TODO: add tests

File: main/src/ca/uwaterloo/flix/language/phase/TypeReconstruction.scala [hunk 1]
```diff
@@ -419,9 +419,17 @@ object TypeReconstruction {
...
       val es = exps.map(visitExp)
-      throw InternalCompilerException(s"Unexpected InvokeConstructor2 call.", loc)
+      val constructorTpe = subst(cvar)
+      val tpe = Type.getFlixType(clazz)
+      val eff = subst(evar)
+      constructorTpe match {
+        case Type.Cst(TypeConstructor.JvmConstructor(constructor), loc) =>
+          TypedAst.Expr.InvokeConstructor(constructor, es, tpe, eff, loc)
+        case _ =>
+          TypedAst.Expr.Error(TypeError.UnresolvedMethod(loc), tpe, eff)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/TypeReconstruction.scala:431]: UnresolvedConstructor?

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala [hunk 2]
```diff
@@ -172,20 +177,48 @@ object TypeReduction {
...
     }
   }
 
-  /**
+    /**
+     * Helper method to retrieve a constructor given its parameter types and the class.
+     * Returns a JavaConstructorResolutionResult containing either the constructor, a list of candidate constructors or
+     * a ConstructorNotFound object. The working process is similar to retrieveMethod and differs in the selection of candidate constructors.
+     */
+  private def retrieveConstructor(clazz: Class[_], ts: List[Type], loc: SourceLocation)(implicit flix: Flix): JavaConstructorResolutionResult = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala:191]: I wonder if we can do some refactor to better share code. What do you think?

**Comment by magnus-madsen** [comment]
In addition to some cleanup of the code, lets add an extensive number of test cases.

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala (hunk 2 above)
Comment (chanattan) [reply main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala:191]: Hm, I also thought about later refactoring.
We may do it if we use match patterns, however method resolution involves a method name which differs from constructor resolution.
We may also consider a JavaResolutionProcess with an option type which could hold either a constructor or a method, depending on the called function the context would de-ambiguize it.
The central point for factorization would be isCandidateMethod/isCandidateConstructor holding different conditions, if we can use also a match pattern here then we could factorize it with a generic JavaResolutionProcess.

File: main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver.scala [hunk 3]
```diff
@@ -282,18 +282,34 @@ object ConstraintSolver {
...
+          case JavaMethodResolutionResult.AmbiguousMethod(methods) => Result.Err(TypeError.AmbiguousMethod(method.name, tpe, tpes, methods, renv, mvar.loc))
+          case JavaMethodResolutionResult.MethodNotFound => Result.Err(TypeError.MethodNotFound(method.name, tpe, tpes, List(), renv, mvar.loc)) // TODO INTEROP: fill in candidate methods
+        }
+      } else {
+        // Otherwise other constraints may still need to be solved.
+        Result.Ok(ResolutionResult(subst0, List(constr0), progress = false))
+      }
+    case TypeConstraint.EqJvmConstructor(cvar, clazz, tpes0, prov) =>
+      // Apply substitution now
+      val tpes = tpes0.map(t => subst0(t))
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver.scala:297]: We tend to prefer `tpes0.map(subst0.apply)`

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala [hunk 4]
```diff
@@ -172,20 +177,48 @@ object TypeReduction {
...
+        // Among candidate constructors if there already exists the one with the exact signature,
+        // we should ignore the rest.
+        val exactConstructor = candidateConstructors
+          .filter(c => // Parameter types correspondance with subtyping
+            (c.getParameterTypes zip ts).forall {
+              case (clazz, tpe) => Type.getFlixType(clazz) == tpe
+            })
+        exactConstructor.length match {
+          case 1 => JavaConstructorResolutionResult.Resolved(Type.Cst(TypeConstructor.JvmConstructor(exactConstructor.head), loc))
+          case _ => JavaConstructorResolutionResult.AmbiguousConstructor(candidateConstructors.toList) // 0 corresponds to no exact constructor, 2 or higher should be impossible in Java
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala:209]: For this and the corresponding "method" case, I suggest we match directly on 0 and then add a default case that gives an `InternalCompilerException`.

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala [hunk 5]
```diff
@@ -146,23 +147,27 @@ object TypeReduction {
         retrieveMethod(clazz, method, ts, loc)
       case Type.Cst(TypeConstructor.Native(clazz), _) =>
         retrieveMethod(clazz, method, ts, loc)
-      case _ => JavaResolutionResult.MethodNotFound
+      case _ => JavaMethodResolutionResult.MethodNotFound
     }
   }
 
+  def lookupConstructor(clazz: Class[_], ts: List[Type], loc: SourceLocation)(implicit flix: Flix): JavaConstructorResolutionResult = {
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala:154]: Docs needed.

**Review by mlutze — COMMENTED** [review]
Good work on this

**Review by magnus-madsen — APPROVED** [review]

**Pull request merged**